#BN-7 feat: Create Roles and Set Permissions

kunda4 opened the pull request:
# PR Description
Admin should be able to assign various users to various roles. There should also be a role for the admin account itself. This will allow only accounts with particular roles permission to peform particular actions.
# Description of tasks that were expected to be completed
List all the tasks and sub-tasks that were assigned to you Make sure you check every completed task.
- [x] admin can create roles
- [x] admin assigns a particular user to a particular role
- [x] the user can only perform actions allowed by their role.
- [x] Document feature using swagger
OPTIONAL FEATURE (CHECK WITH MANAGER)
- [x] Each role should have a list of permissions attached to them.
- [x] Admin should be able to add and remove permissions from particular roles.
# How this may be tested?
Please describe the tests that you ran to verify your functionalities. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.
1. run `npm i` to install all required dependencies
2. run `npm run dev`
3. step  use postman to test endpoint or swagger
4. step 4...
# Screenshots (If appropriate/applicable)
Provide screenshots related to your PR in case it's appropriate and applicable
# Please check this Checklist before you submit your PR:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] My code generates no warnings
- [x] There are no vulnerabilities

Honore123 committed:
fix(resolve): merge conflicts

Honore123 committed:
fix(resolve): branch conflicts

UmubyeyiEvelyne committed:
chore(#BN-9: Add two factor authenrtication for vendors): Added create user and created a secret token that will be used to generate an OTP sent to the user via SMS

UmubyeyiEvelyne committed:
feat(#BN-9: Two factor authentication for vendors): I added sending OTP message to the user upon successful registration using twilio

Honore123 committed:
fix(rename): rename database
name database based on running environment
[done #26]

Honore123 committed:
feat(Makefile): create Makefile
reduce docker commands
[done #26]

Honore123 committed:
feat(Makefile): create Makefile
reduce docker commands
[done #26]

Honore123 committed:
feat(Makefile): create Makefile
reduce docker commands
[done #26]

Honore123 committed:
feat(Makefile): create Makefile
reduce docker commands
[done #26]

UmubyeyiEvelyne committed:
feat(#BN-9: Two factor authentication for vendors): I added endpoints for sending OTP message to user's email, and another one for validating OTP submitted by the user
- I added documentation for these two endpoints

Commit:
#3 chore: create a buyer sign up , send a confirmation email and sign up with a social account

Commit:
fixed the issue in the response message and response file

UmubyeyiEvelyne committed:
feat(#BN-9: Two factor authentication for vendors): added tests for validate OTP endpoint

UmubyeyiEvelyne committed:
feat(#BN-9: Two factor authentication for vendors): changed the dev script in package.json

Commit:
added translation message

Commit:
added some changes to the response message

Commit:
added some changes to the response message

Commit:
added some changes to the response message

UmubyeyiEvelyne committed:
feat(#BN-9: Two factor authentication for vendors): resolved eslint conflicts in orderItems.js

Commit:
fix-merge

Commit:
fix-conflict

Commit:
fix-conflict

Commit:
fix-conflict

Commit:
fix-conflict

Commit:
fix-conflict

Commit:
fix-conflict

Commit:
feat(Create Role and Set Permission): Create Role and Set Permission feature
-modify swagger documentation
-set middleware for isAdmin, isSeller and isBuyer and checkPermission
-work on translation of response message
[Finishes #7]

Commit:
feat(Create Role and Set Permission): Create Role and Set Permission feature
-modify swagger documentation
-set middleware for isAdmin, isSeller and isBuyer and checkPermission
-work on translation of response message
[Finishes #7]

chrisostome01 commented on src/controllers/permController.js:
```diff
@@ -0,0 +1,201 @@
+                });
+            } else {              
+                await permission.update(
+                    { permName, description },
```
Better to have permName as permssionName

chrisostome01 commented on src/controllers/permController.js:
Please consider formatting this file eslint prettier

chrisostome01 replied:
permssionController would be a good choice for this fileName and it would be more descriptive enough

chrisostome01 commented on src/controllers/permController.js:
```diff
@@ -0,0 +1,201 @@
+            const newPermission = await permission.create({permName, description})
+            return res
+        .status(201)
+        .json({ message: req.t('permission_create_message'), data: newPermission });
```
remember to include you statusCode in the response too

chrisostome01 commented on src/controllers/permController.js:
```diff
@@ -0,0 +1,201 @@
+        .json({ message: req.t('permission_create_message'), data: newPermission });
+        } 
+        catch (err) {
+            console.log('error ' + err);
```
Remove this console

chrisostome01 commented on src/controllers/permController.js:
```diff
@@ -0,0 +1,201 @@
+import {permission,rolepermission} from "../database/models";
+    
+    const createNewPermission = async(req, res)=> {
+    const { permName, description } = req.body; 
```
PermName, should be meaningful.

chrisostome01 commented on src/controllers/permController.js:
```diff
@@ -0,0 +1,201 @@
+    const getAllPermission = async(req, res)=> {
+        try {
+            const permissions = await permission.findAll();
+            res.status(200).json({ statusCode: 200, data: permissions });
```
Message is missing in your response

chrisostome01 commented on src/controllers/permController.js:
```diff
@@ -0,0 +1,201 @@
+                where: { id },
+            });
+            if (permissions) {
+                return res.status(200).json({ statusCode: 200, data: permissions });
```
Message is missing here

chrisostome01 commented on src/controllers/permController.js:
```diff
@@ -0,0 +1,201 @@
+            } else {
+                res.status(404).json({
+                    statusCode: 404,
+                    Message: req.t('id_unexist_message'),
```
Stay consistent `Message` is completely different from `message`

chrisostome01 commented on src/controllers/roleController.js:
```diff
@@ -0,0 +1,207 @@
+          }
+        try {
+            const userRoles = await userRole.create(roleData)
+            return res.status(201).json({ message: req.t('set-role-message'), response: userRoles }) 
```
snake case would be good instead `set-role-message` you can use `set_role_message`

chrisostome01 commented on src/controllers/roleController.js:
```diff
@@ -0,0 +1,207 @@
+                    .json({ statusCode: 201, message: req.t('update_message'),data:UserWithRole });
+            }
+        } catch (error) {
+            console.error(error);
```
remove this console

chrisostome01 commented on src/controllers/roleController.js:
```diff
@@ -0,0 +1,207 @@
+        } catch (error) {
+            console.error(error);
+            return res
+                .status(500)
+                .json({ statusCode: 400, error: "Server error" });
```
status code are deferent...

chrisostome01 commented on src/controllers/userController.js:
```diff
@@ -73,9 +73,9 @@ export async function loginUser(req, res) {
+    firstName, lastName, email, password, phone, roleId,
   } = req.body;
-  if (!firstName || !lastName || !email || !password || !phone || !role) {
+  if (!firstName || !lastName || !email || !password || !phone || !roleId) {
```
Use Joi to validate all of those inputs

chrisostome01 commented on src/controllers/permController.js:
```diff
@@ -0,0 +1,201 @@
+                });
+            }
+        } catch (error) {
+            return res.status(500).json({ error: error.message })
```
Use message here instead of error to improve consistency

chrisostome01 reviewed (changes requested):
# Well done @kunda4 the codes are really well structured however here are a couple of things that we need to work on.

- Improve consistency in your code structure: some lines of code are written differently but in a different way but the same context. `message` and `Message` or `error` instead of `message`
- You can generate your own seeds instead of reusing the same file.
- Please make sure you are validating all of the data you are receiving 
- Please format the code for more structured codes

chrisostome01 commented on Makefile:
```diff
@@ -19,4 +19,8 @@ migrate-rollback:
+	docker exec e_commerce-node-app-1 nom run seed-down
+=======
+	docker exec e_commerce-node-app-1 nom run seed-down
+>>>>>>> a84c243cf595935bb3c658307b81e93305ae41fb
```
I those conflicts are not supposed to be here..?

chrisostome01 commented on package.json:
```diff
@@ -4,9 +4,11 @@
   "description": "",
   "main": "index.js",
   "scripts": {
+    "make-model": "npx sequelize-cli model:generate --name users --attributes firstName:string,lastName:string,email:string, password:string, role:string",
```
Remove those demo data `users firstName:string,lastName:string,email:string, password:string, role:string`

chrisostome01 commented on package.json:
```diff
@@ -4,9 +4,11 @@
     "migrate": "npx sequelize-cli db:migrate",
     "seed": "npx sequelize-cli db:seed:all",
     "seed-down": "npx sequelize-cli db:seed:undo:all",
+    "make-seed": "npx sequelize-cli seed:generate --name demo-users",
```
remove also this `--name demo-users` this should be someone else's inputs

Commit:
resolve conversation and fx-conflict

Commit:
fix(bug): fix user validation
- Create newUser.validation middleware
- Add newUser.validation middleware in routes
- Remove console messages
- Improve consistency in my code structure
- Format code with Prettier eslint
[Finishes #7]

Commit:
fix(bug): fix user validation
- Create newUser.validation middleware
- Add newUser.validation middleware in routes
- Remove console messages
- Improve consistency in my code structure
- Format code with Prettier eslint
[Finishes #7]

Commit:
Merge branch 'ft-Create-Roles-Route-bn-7' of https://github.com/atlp-rwanda/ecommerce-app-cogito-bn into ft-Create-Roles-Route-bn-7

Honore123 reviewed (approved):
🚀

chrisostome01 reviewed (approved):
# LGTM
